src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -646,21 +647,33 @@ static int am_recv_event(struct fi_cq_tagged_entry *wc, MPIR_Request * rreq)
 
             break;
         case MPIDI_AMTYPE_PIPELINE:
-            p_data = (char *) wc->buf + sizeof(*am_hdr) + am_hdr->am_hdr_sz;
+            p_data = (char *) orig_buf + sizeof(*am_hdr) + am_hdr->am_hdr_sz;
             mpi_errno = MPIDI_OFI_handle_pipeline(am_hdr, am_hdr + 1, p_data);
             MPIR_ERR_CHECK(mpi_errno);
             break;
 
         case MPIDI_AMTYPE_RDMA_READ:
-            /* buffer always copied together (there is no payload, just LMT header) */
-            p_data = (char *) am_hdr + sizeof(*am_hdr) + am_hdr->am_hdr_sz;
-            mpi_errno = MPIDI_OFI_handle_rdma_read(am_hdr, am_hdr + 1,
-                                                   (MPIDI_OFI_lmt_msg_payload_t *) p_data);
-
-            MPIR_ERR_CHECK(mpi_errno);
+            {
+                /* buffer always copied together (there is no payload, just LMT header) */
+#if NEEDS_STRICT_ALIGNMENT
+                MPIDI_OFI_lmt_msg_payload_t temp_rdma_payload;
+                if (!need_alignment_check) {
+                    memcpy(&temp_rdma_payload,
+                           (char *) orig_buf + sizeof(*am_hdr) + am_hdr->am_hdr_sz,
+                           sizeof(MPIDI_OFI_lmt_msg_payload_t));
```

Comment:
Oh, `MPIDI_OFI_lmt_msg_payload_t` is not the payload, it is an (extended) header that describes the payload. I find `temp_rdma_payload` misleading. I think you can simply name the variable `temp`. It is a temp variable used within 4 statements, I think it is cleaner to just use `temp` than to be misled by the name (which doesn't really help anything here).

Reply:
the datatype is called `MPIDI_OFI_lmt_msg_payload_t`, so I think we can change it to `temp_rdma_lmt_msg`